compiler/modules/nback/function.bal

```diff
@@ -0,0 +1,186 @@
+        // runtime treats nArgs as unsigned64 (note llvm don't distinguish signed and unsigned integer types)
+        nArgs = builder.iArithmeticWrap("add", constInt(scaffold, requiredArgCount), restArgCount);
+        restArgs = restArgArray;
+    }
+    else {
+        nArgs = constInt(scaffold, requiredArgCount);
+        restArgs = ();
+    }
+    llvm:PointerValue uniformArgArray = <llvm:PointerValue>buildRuntimeFunctionCall(builder, scaffold,
+                                                                                    constructUniformArgArrayFunction, [nArgs]);
+    foreach int i in 0 ..< uniformArgs.length() {
+        builder.store(uniformArgs[i], builder.getElementPtr(uniformArgArray, [constInt(scaffold, i)], "inbounds"));
+    }
+    if restArgs != () {
+        buildVoidRuntimeFunctionCall(builder, scaffold, addRestArgsToUniformArgsFunction,
+                                     [uniformArgArray, restArgs, constInt(scaffold, uniformArgs.length())]);
+    }
+    llvm:PointerValue funcPtr = builder.getElementPtr(funcValuePtr, [constIndex(scaffold, 0),
+                                                                     constIndex(scaffold, 1)],
+                                                      "inbounds");
+    llvm:PointerValue funcDesc = <llvm:PointerValue>builder.load(builder.getElementPtr(funcValuePtr,
+                                                                                       [constIndex(scaffold, 0),
+                                                                                        constIndex(scaffold, 0)],
+                                                                                       "inbounds"));
+    llvm:PointerValue uniformFuncPtr = builder.getElementPtr(funcDesc, [constIndex(scaffold, 0),
+                                                                        constIndex(scaffold, 1)],
+                                                             "inbounds");
+    llvm:Value? returnVal = buildFunctionCall(builder, scaffold, <llvm:PointerValue>builder.load(uniformFuncPtr),
+                                              [builder.load(funcPtr), uniformArgArray, nArgs]);
+    if returnVal !is llvm:PointerValue {
+        panic err:impossible("uniform call must return a tagged pointer");
+    }
+    builder.store(exactReturnValue(builder, scaffold, semTypeRetRepr(returnType), returnVal),
+                  scaffold.address(insn.result));
+    builder.br(afterCall);
+}
+
+function exactReturnValue(llvm:Builder builder, Scaffold scaffold, RetRepr retRepr, llvm:PointerValue returnVal) returns llvm:Value {
+    if retRepr is BooleanRepr {
+        return buildUntagBoolean(builder, returnVal);
+    }
+    else if retRepr is IntRepr {
+        return buildUntagInt(builder, scaffold, returnVal);
+    }
+    else if retRepr is FloatRepr {
+        return buildUntagFloat(builder, scaffold, returnVal);
+    }
+    return returnVal;
+}
+
+function uniformRepr(t:SemType ty) returns TaggedRepr {
```

Comment:
This doesn't feel like it's function specific. There's similar functionality needed for structure members, which are usually tagged.

Reply:
Fixed